File: nl.naturalis.nba.client.php/model/TaxonomicRank.php
<?php

namespace nl\naturalis\nba\client\php\model;

final class TaxonomicRank {

	private static $kingdom;
	private static $phylum;
	private static $classis;
	private static $order;
	private static $superFamily;
	private static $family;
	private static $tribe;
	private static $genus;
	private static $subgenus;
	private static $species;
	private static $subspecies;

	public static function KINGDOM()
	{
		if (self::$kingdom === null)
			self::$kingdom = new TaxonomicRank("kingdom", "regnum");
		return self::$kingdom;
	}

	public static function PHYLUM()
	{
		if (self::$phylum === null)
			self::$phylum = new TaxonomicRank("phylum");
		return self::$phylum;
	}

	public static function CLASSIS()
	{
		if (self::$classis === null)
			self::$classis = new TaxonomicRank("class", "classis");
		return self::$classis;
	}

	public static function ORDER()
	{
		if (self::$order === null)
			self::$order = new TaxonomicRank("order", "ordo");
		return self::$order;
	}

	public static function SUPER_FAMILY()
	{
		if (self::$superFamily === null)
			self::$superFamily = new TaxonomicRank("superfamily", "suprafamilia");
		return self::$superFamily;
	}

	public static function FAMILY()
	{
		if (self::$family === null)
			self::$family = new TaxonomicRank("family", "familia");
		return self::$family;
	}

	public static function TRIBE()
	{
		if (self::$tribe === null)
			self::$tribe = new TaxonomicRank("tribe", "tribus");
		return self::$tribe;
	}

	public static function GENUS()
	{
		if (self::$genus === null)
			self::$genus = new TaxonomicRank("genus");
		return self::$genus;
	}

	public static function SUBGENUS()
	{
		if (self::$subgenus === null)
			self::$subgenus = new TaxonomicRank("subgenus");
		return self::$subgenus;
	}

	public static function SPECIES()
	{
		if (self::$species === null)
			self::$species = new TaxonomicRank("species");
		return self::$species;
	}

	public static function SUBSPECIES()
	{
		if (self::$subspecies === null)
			self::$subspecies = new TaxonomicRank("subspecies");
		return self::$subspecies;
	}

	private final $englishName;
	private final $latinName;

	private function __construct(String $englishName, String $latinName)
	{
		$this->englishName = $englishName;
		if (isset($latinName))
			$this->latinName = $latinName;
		else
			$this->latinName = $englishName;
	}

	public function getEnglishName()
	{
		return $this->englishName;
	}

	public function getLatinName()
	{
		return $this->latinName;
	}

	public function __toString()
	{
		return $this->englishName;
	}
}

File: nl.naturalis.nba.api/src/main/java/nl/naturalis/nba/api/model/TaxonomicRank.java
package nl.naturalis.nba.api.model;

public enum TaxonomicRank implements INbaModelObject
{

  DOMAIN("domain", "regio"),
	KINGDOM("kingdom", "regnum"),
	SUBKINGDOM("subkingdom"),
	PHYLUM("phylum"),
	SUBPHYLUM("subphylum"),
	SUPERCLASS("superclass"),
	CLASS("class", "classis"),
	SUBCLASS("subclass"),
	SUPERORDER("superorder"),
	ORDER("order", "ordo"),
	SUBORDER("suborder"),
	INFRA_ORDER("infraorder"),
	SUPER_FAMILY("superfamily", "suprafamilia"),
	FAMILY("family", "familia"),
	SUBFAMILY("subfamily"),
	TRIBE("tribe", "tribus"),
	SUBTRIBE("subtribe"),
	GENUS("genus"),
	SUBGENUS("subgenus"),
	SPECIES("species"),
	SUBSPECIES("subspecies");

	private final String englishName;
	private final String latinName;

	private TaxonomicRank(String englishName, String latinName)
	{
		this.englishName = englishName;
		this.latinName = latinName;
	}

	private TaxonomicRank(String latinName)
	{
		this.englishName = latinName;
		this.latinName = latinName;
	}

	public String getEnglishName()
	{
		return englishName;
	}

	public String getLatinName()
	{
		return latinName;
	}

	public String toString()
	{
		return englishName;
	}

	public static TaxonomicRank parse(String name)
	{
		if (name == null) {
			return null;
		}
		name = name.toLowerCase();
		if (name.equals(KINGDOM.englishName) || name.equals(KINGDOM.latinName)) {
			return KINGDOM;
		}
    if (name.equals(SUBKINGDOM.englishName)) {
      return SUBKINGDOM;
    }
		if (name.equals(PHYLUM.latinName)) {
			return PHYLUM;
		}
    if (name.equals(SUBPHYLUM.englishName)) {
      return SUBPHYLUM;
    }
    if (name.equals(SUPERCLASS.englishName)) {
      return SUPERCLASS;
    }
		if (name.equals(CLASS.englishName) || name.equals(CLASS.latinName)) {
			return CLASS;
		}
    if (name.equals(SUBCLASS.englishName)) {
      return SUBCLASS;
    }
    if (name.equals(SUPERORDER.englishName)) {
      return SUPERORDER;
    }
		if (name.equals(ORDER.englishName) || name.equals(ORDER.latinName)) {
			return ORDER;
		}
    if (name.equals(SUBORDER.englishName)) {
      return SUBORDER;
    }
    if (name.equals(INFRA_ORDER.englishName)) {
      return INFRA_ORDER;
    }
    if (name.equals(SUPER_FAMILY.englishName) || name.equals(SUPER_FAMILY.latinName)) {
			return SUPER_FAMILY;
		}
		if (name.equals(FAMILY.englishName) || name.equals(FAMILY.latinName)) {
			return FAMILY;
		}
    if (name.equals(SUBFAMILY.englishName)) {
      return SUBFAMILY;
    }
		if (name.equals(TRIBE.englishName) || name.equals(TRIBE.latinName)) {
			return TRIBE;
		}
    if (name.equals(SUBTRIBE.englishName)) {
      return SUBTRIBE;
    }
		if (name.equals(GENUS.latinName)) {
			return GENUS;
		}
		if (name.equals(SUBGENUS.latinName)) {
			return SUBGENUS;
		}
		if (name.equals(SPECIES.latinName)) {
			return SPECIES;
		}
		if (name.equals("species epithet") || name.equals("speciesepithet")) {
			return SPECIES;
		}
		if (name.equals("specific epithet") || name.equals("specificepithet")) {
			return SPECIES;
		}
		if (name.equals(SUBSPECIES.latinName)) {
			return SUBSPECIES;
		}
		if (name.equals("subspecies epithet") || name.equals("subspeciesepithet")) {
			return SUBSPECIES;
		}
		if (name.equals("infraspecific epithet") || name.equals("infraspecificepithet")) {
			return SUBSPECIES;
		}
		return null;
	}

}